oauth2_http/java/com/google/auth/oauth2/UserAuthorizer.java

```diff
@@ -285,61 +295,36 @@ public UserCredentials getCredentialsFromCode(String code, URI baseUri) throws I
-    HttpRequest tokenRequest =
-        requestFactory.buildPostRequest(new GenericUrl(tokenServerUri), tokenContent);
-    tokenRequest.setParser(new JsonObjectParser(OAuth2Utils.JSON_FACTORY));
-
-    HttpResponse tokenResponse = tokenRequest.execute();
-
-    GenericJson parsedTokens = tokenResponse.parseAs(GenericJson.class);
-    String accessTokenValue =
-        OAuth2Utils.validateString(parsedTokens, "access_token", FETCH_TOKEN_ERROR);
-    int expiresInSecs = OAuth2Utils.validateInt32(parsedTokens, "expires_in", FETCH_TOKEN_ERROR);
-    Date expirationTime = new Date(new Date().getTime() + expiresInSecs * 1000);
-    String scopes =
-        OAuth2Utils.validateOptionalString(
-            parsedTokens, OAuth2Utils.TOKEN_RESPONSE_SCOPE, FETCH_TOKEN_ERROR);
-    AccessToken accessToken =
-        AccessToken.newBuilder()
-            .setExpirationTime(expirationTime)
-            .setTokenValue(accessTokenValue)
-            .setScopes(scopes)
-            .build();
-    String refreshToken =
-        OAuth2Utils.validateOptionalString(parsedTokens, "refresh_token", FETCH_TOKEN_ERROR);
-
+    TokenResponseWithConfig tokenResponseWithConfig =
+        getCredentialsFromCodeInternal(code, baseUri, additionalParameters);
     return UserCredentials.newBuilder()
-        .setClientId(clientId.getClientId())
-        .setClientSecret(clientId.getClientSecret())
-        .setRefreshToken(refreshToken)
-        .setAccessToken(accessToken)
-        .setHttpTransportFactory(transportFactory)
-        .setTokenServerUri(tokenServerUri)
+        .setClientId(tokenResponseWithConfig.getClientId())
+        .setClientSecret(tokenResponseWithConfig.getClientSecret())
+        .setAccessToken(tokenResponseWithConfig.getAccessToken())
+        .setRefreshToken(tokenResponseWithConfig.getRefreshToken())
+        .setHttpTransportFactory(tokenResponseWithConfig.getHttpTransportFactory())
+        .setTokenServerUri(tokenResponseWithConfig.getTokenServerUri())
         .build();
   }
 
+  /**
+   * Handles OAuth2 authorization code exchange and returns a {@link TokenResponseWithConfig} object
+   * containing the tokens and configuration details.
+   *
+   * @param code The authorization code received from the OAuth2 authorization server.
+   * @param callbackUri The URI to which the authorization server redirected the user after granting
+   *     authorization.
+   * @param additionalParameters Additional parameters to include in the token exchange request.
+   * @return A {@link TokenResponseWithConfig} object containing the access token, refresh token (if
+   *     granted), and configuration details used in the OAuth flow.
+   * @throws IOException If an error occurs during the token exchange process.
+   */
+  public TokenResponseWithConfig getTokenResponseFromAuthCodeExchange(
+      String code, URI callbackUri, @Nullable Map<String, String> additionalParameters)
```

Comment:
qq, can we make additionalParameters required to be non-null? If there are no additionalParameters needed, the user can just pass in an empty map. We wouldn't need the null check below in `getCredentialsFromCodeInternal` for additionalParameters

Reply:
I can remove `@Nullable` but I can't remove the null check. This is existing code that is used by getCredentialsFromCode. Removing the null check would be a breaking change